### PR TITLE
Add support for errors in zebra_test::Transcript

### LINKED ISSUE #676: Model errors in service transcripts.
Right now the service `Transcript` can't check that requests fail, only that they succeed, because the transcript is a list of request/response pairs, not a list of request/result pairs. This could probably be fixed by changing
```rust
pub struct Transcript<R, S, I>
where
    I: Iterator<Item = (R, S)>,
```
to
```rust
pub struct Transcript<R, S, E, I>
where
    I: Iterator<Item = (R, Result<S, E>)>,
```
so that the transcript is of responses or errors, but I'm not sure what the right way to change the check function is.

Currently it's
```rust
impl<R, S, I> Transcript<R, S, I>
where
    I: Iterator<Item = (R, S)>,
    R: Debug,
    S: Debug + Eq,
{
    pub async fn check<C>(mut self, mut to_check: C) -> Result<(), Report>
    where
        C: Service<R, Response = S>,
        C::Error: Debug,
    { /* ... */ }
}
```
but the obvious change to
```rust
impl<R, S, E, I> Transcript<R, S, E, I>
where
    I: Iterator<Item = (R, S)>,
    R: Debug,
    S: Debug + Eq,
    E: Debug + Eq,
{
    pub async fn check<C>(mut self, mut to_check: C) -> Result<(), Report>
    where
        C: Service<R, Response = S, Error = E>,
    { /* ... */ }
}
```
isn't right because most errors aren't `Eq`, it's not clear what that would mean anyways, and many errors are extremely inconvenient to construct outside of their normal context.

If we don't care about testing **specific** facts about an error, but just that an error occurred, then rather than adding an extra error type `E`, we could instead change the `Transcript` definition to
```rust
pub struct Transcript<R, S, I>
where
    I: Iterator<Item = (R, Option<S>)>,
```
and use `Result::ok` on the result of the service call to check for the *presence* of an error rather than the details of the error.

### PR DESCRIPTION
**Pull request by yaahc**
closes https://github.com/ZcashFoundation/zebra/issues/676

### CODE REVIEW

File: zebra-state/tests/basic.rs
Comment (yaahc) [inline zebra-state/tests/basic.rs]: We will probably want to export this alias in `zebra_test`, or maybe make it an associated type alias with the `Transcript`.

File: zebra-test/src/transcript.rs
Comment (yaahc) [inline zebra-test/src/transcript.rs]: These two arms need to be filled out but I rushed because it was meeting time.

File: zebra-test/src/transcript.rs
Comment (yaahc) [inline zebra-test/src/transcript.rs]: This is what I came up with for handling errors when using transcripts as a mock. LMK what you think @hdevalence

File: zebra-test/src/transcript.rs
Comment (yaahc) [inline zebra-test/src/transcript.rs]: I think this is cool and want other people to notice and appreciate it too

File: zebra-test/tests/transcript.rs
Comment (yaahc) [inline zebra-test/tests/transcript.rs]: My guess is that we're going to want to just define tests in the test fn so we can use `let` and type inference.

File: zebra-state/tests/basic.rs
Comment (hdevalence) [reply zebra-state/tests/basic.rs]: Both aliases or just one?

File: zebra-test/tests/transcript.rs
Comment (yaahc) [reply zebra-test/tests/transcript.rs]: type inference was some nightmare shit, it looks like using consts / statics, manually specified types, and fn pointers gives the most ergonomic interface.

File: zebra-state/tests/basic.rs
Comment (hdevalence) [inline zebra-state/tests/basic.rs:47]: I think it would be good to extend this transcript with an error case to see the error ergonomics in practice.

File: zebra-test/src/transcript.rs
Comment (hdevalence) [inline zebra-test/src/transcript.rs]: Because the transcript items are requests and responses to the service, I don't think it makes sense to check the result of poll_ready against the transcript.  We should require that the service does not fail during the execution of the transcript. Once a service returns an error from `poll_ready` it's assumed to be failed permanently (this is some possibly documented part of the `Service` contract), so there's not much information to be gained by testing failures of the service itself, rather than failures of individual requests.

File: zebra-test/src/transcript.rs
Comment (hdevalence) [reply zebra-test/src/transcript.rs]: I don't think I understand why the guide is a static str rather than part of the `Transcript` documentation.

File: zebra-test/src/transcript.rs
Comment (hdevalence) [inline zebra-test/src/transcript.rs]: I'm not sure I understand this part.  Is this supposed to handle the "dual intent" usecase, where a `Transcript` is supposed to be able to be both a `Service` impl itself as well as to check an existing `Service`? Having to write this every time seems a little bit unergonomic but I'm not yet sure what a better alternative is.

File: zebra-test/src/transcript.rs
Comment (yaahc) [reply zebra-test/src/transcript.rs]: Because this is runtime information that is output as a custom section in the error report when someone writes a test that is clearly bugged.

File: zebra-state/tests/basic.rs
Comment (yaahc) [reply zebra-state/tests/basic.rs:47]: I didn't do it for this example but I did add an example using the closures to the `zebra-tests/tests/transcript.rs` file

File: zebra-test/src/transcript.rs
Comment (yaahc) [reply zebra-test/src/transcript.rs]: Yea, I was drafting a comment about how the `MockError` stuff was bad and asking for help on how to fix it when I realized I should just use `Option<BoxError>` as the input instead. You barely reviewed it before I pushed the new version.

File: zebra-state/tests/basic.rs
Comment (yaahc) [reply zebra-state/tests/basic.rs:47]: Ended up doing that here https://github.com/ZcashFoundation/zebra/pull/678/files#diff-9bc388fb6fb0b6f6cadd80bbd760b1c7R57

File: zebra-state/tests/basic.rs
Comment (yaahc) [reply zebra-state/tests/basic.rs]: Just the one seems best.

File: zebra-test/src/transcript.rs
Comment (yaahc) [inline zebra-test/src/transcript.rs]: This is the first time I've ever had a good excuse to use `catch_unwind`

**Comment by yaahc** [comment]
This was a fun PR :D

**Review by oxarbitrage — DISMISSED** [review]
It looks good to me, good test coverage and the documentation helps to understand the whole idea. It will be nice if @hdevalence can also approve before merging. Needed for a re implementation of https://github.com/ZcashFoundation/zebra/pull/675 using transcript.

**Comment by yaahc** [comment]
@oxarbitrage if you want you can try rebasing your branch ontop of `transcript-errors` for now so you can move ahead with the work in #675. Hopefully we wont need any big changes before this merges so you won't have to deal with merge conflicts.

File: zebra-test/tests/transcript.rs
Comment (hdevalence) [inline zebra-test/tests/transcript.rs]: This is a lot more unwieldy than the non-error case, so I think this is a signal that we're not defining the problem and modeling the solution correctly.  It's worth re-evaluating the problem space before continuing here.

**Review by dconnolly — DISMISSED** [review]
lgtm 🎈

**Comment by yaahc** [comment]
@dconnolly I think henry wants to revisit the design so I'm gonna hold off on merging until he posts that design comment he mentioned above

**Review by teor2345 — APPROVED** [review]
Looks great!

It looks like this PR makes improvements to the design, and it has some tests.

So I'm going to suggest that we merge, and do any follow-ups in future PRs or issues.

**Pull request merged**
